[Chore] Submodule pattern updates to TreeStore

dylanlott opened the pull request:
## Description

This PR updates the TreeStore submodule to use the latest submodule pattern updates.

## Issue

Part of the work of implementing #562 

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- Makes the TreeStore take proper advantage of the new `Submodule` interface 

## Testing

- [x] `make develop_test`; if any code changes were made
- [x] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [x] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

<!-- REMOVE this comment block after following the instructions
 If you added additional tests or infrastructure, describe it here.
 Bonus points for images and videos or gifs.
-->

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [x] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

h5law commented on persistence/module.go:
```diff
@@ -106,9 +106,14 @@ func (*persistenceModule) Create(bus modules.Bus, options ...modules.ModuleOptio
-		return nil, err
+		return nil, fmt.Errorf("failed to create TreeStoreModule: %w", err)
+	}
+	treeStoreModule, ok := treeModule.(modules.TreeStoreModule)
```
Can you explain why this is needed? does not `trees.Create` return `modules.TreeStoreModule`?

h5law commented on persistence/trees/module.go:
```diff
@@ -3,19 +3,26 @@ package trees
-	m := &treeStore{}
+var _ modules.Module = &TreeStore{}
+
+func (*TreeStore) Create(bus modules.Bus, options ...modules.ModuleOption) (modules.Module, error) {
```
Is there a reason to not return `modules.TreeStoreModule` here? Now submodules work properly?

h5law reviewed (changes requested):
I am unsure why you have moved away from `TreeStore` being a submodule but instead making it a full module. Other than that the changes are good I have actually addressed this in my `ibc/host_module` branch.

👍🏼 for the new mockgen line ;)

h5law commented on persistence/trees/module.go:
```diff
@@ -24,35 +31,52 @@ func (*treeStore) Create(bus modules.Bus, options ...modules.TreeStoreOption) (m
-func (t *treeStore) setupTrees() error {
-	if t.treeStoreDir == ":memory:" {
+func (t *TreeStore) GetModuleName() string  { return modules.TreeStoreModuleName }
+func (t *TreeStore) Start() error           { return nil }
```
These are noop, `TreeStore` really seems like it should not be a `Module` but a `Submodule` like it was previously

h5law commented on persistence/trees/module.go:
```diff
@@ -24,35 +31,52 @@ func (*treeStore) Create(bus modules.Bus, options ...modules.TreeStoreOption) (m
+func (t *TreeStore) GetModuleName() string  { return modules.TreeStoreModuleName }
+func (t *TreeStore) Start() error           { return nil }
+func (t *TreeStore) Stop() error            { return nil }
+func (t *TreeStore) GetBus() modules.Bus    { return t.Bus }
```
I dont think you need these as you are including the `base_module.IntegrableModule` in the `TreeStore` struct. Only one you **need** here is `GetModuleName()`

dylanlott commented:
> I am unsure why you have moved away from `TreeStore` being a submodule but instead making it a full module. Other than that the changes are good I have actually addressed this in my `ibc/host_module` branch.
> 
> 👍🏼 for the new mockgen line ;)

Updated to use the Submodule type as the title suggests, I had a bad merge somewhere that I didn't catch. Fixed now 👍

Olshansk commented on persistence/module.go:
```diff
@@ -117,7 +118,7 @@ func (*persistenceModule) Create(bus modules.Bus, options ...modules.ModuleOptio
 	m.blockStore = blockStore
 	m.txIndexer = txIndexer
-	m.stateTrees = treeModule
+	m.stateTrees = treeStore
```
Should we rename `stateTrees` to `treeStore`

Olshansk commented on persistence/trees/trees.go:
```diff
@@ -71,26 +71,29 @@ type stateTree struct {
-	logger       *modules.Logger
-	treeStoreDir string
+	logger *modules.Logger
+	Bus    modules.Bus
```
This feels like a code smell. I don't think `Bus` should ever be an exposed structure. It's the "bus" that everyone has access to internally

Olshansk commented on persistence/trees/trees.go:
```diff
@@ -71,26 +71,29 @@ type stateTree struct {
 // of the underlying trees by utilizing the lazy loading
 // functionality provided by the underlying smt library.
-type treeStore struct {
+type TreeStore struct {
```
Given that we have an interface `TreeStoreModule`, we shouldn't need to publically expose the implementation as well. Otherwise, feels like we're not encapsulating/defining logic correctly.

1. Why not update TreeStoreModule?
2. What needs to happen to not expose this?

dylanlott replied:
This is necessary for testing. If we don't do this, then we can't access the right methods, and if we keep the tests in the same package as trees, then we get import errors because we haven't ironed out some of the dependencies.

Olshansk replied:
In the future, I think that if we're making a design decision solely for testing, we should add a `TECHDEBT` comment that says something along the lines of "exposed for testing purposes.

---

> This is necessary for testing. If we don't do this, then we can't access the right methods

Can you look at `ConsensusDebugModule` and expose the functions we need in a testing interface with a `TECHDEBT(#880)` comment above it?

> then we get import errors because we haven't ironed out some of the dependencies.

Can you add details on the import errors in #880? I'm guessing they're related. cc @bryanchriswhite

bryanchriswhite replied:
If testing is the only place where this would be imported you can add a file to this package which includes the `//go:build test` build tag and exports it for usage in tests:

```
type TreeStore = treeStore

func (ts *treeStore) SomeTestHelperMethod(...) { ... }
```

See [p2p/raintree/testutil.go](https://github.com/pokt-network/pocket/blob/main/p2p/raintree/testutil.go) or [p2p/testutil.go](https://github.com/pokt-network/pocket/blob/main/p2p/testutil.go) for examples.

bryanchriswhite replied:
> Can you add details on the import errors in #880? I'm guessing they're related. cc @bryanchriswhite

:100: :pray:

dylanlott replied:
@Olshansk I added a techdebt comment linking to #880 and will include the test tag changes in a future PR that incorporates the tests. The capitalization is in this PR because I split the branch into multiple PRs after I had written the tests.

@bryanchriswhite I'll include the tag and custom struct technique in the [TestUtils doc](https://www.notion.so/pocketnetwork/Testutils-9cba9010e18447248e9daa8a3b87e3f2) as well, because it's mentioned but not detailed and it's a smart trick to get around import errors for now until we fix the deeper issue, and might still be useful later on anyway.

Olshansk replied:
@dylanlott I created a branch #920 and ran the tests and everything worked out fine.

Where are the tests that are blocking this?

dylanlott replied:
They're in a WIP PR that was based off of the branch that I split up to make these first three commits, so this will work right now, and I'll use the approach @bryanchriswhite suggested to adjust the package scope for the names later, that works fine by me.

Olshansk commented on persistence/trees/trees.go:
```diff
@@ -71,26 +71,29 @@ type stateTree struct {
-	treeStoreDir string
+	logger *modules.Logger
+	Bus    modules.Bus
+	TXI    indexer.TxIndexer
```
Ditto regarding exposing internal implementation

bryanchriswhite commented on persistence/module.go:
```diff
@@ -239,7 +239,7 @@ func (m *persistenceModule) GetTxIndexer() indexer.TxIndexer {
 }
 
 func (m *persistenceModule) GetTreeStore() modules.TreeStoreModule {
-	return m.stateTrees
+	return m.treeStore
 }
```
This method becomes redundant now that `treeStore` is dependency injectable.

Consumers of this method should retrieve it from the bus/registry e.g. `bus.GetTreeStore()`. Until #880 is resolved you won't be able to implement `Bus#GetTreeStore()`, which is the same thing that happened in the `PeerstoreProvider` submodule refactor (see: #895).

You can work around this in a similar way, using an unexported temporary helper with does the `bus.GetModuleRegistry().GetModule(...)` call. See [`p2p#getPeerstoreProvider`](https://github.com/pokt-network/pocket/blob/refactor/p2p-submodules/p2p/module.go#L530) as an example.

bryanchriswhite commented on persistence/module.go:
```diff
@@ -41,9 +41,9 @@ type persistenceModule struct {
+	// treeStore manages all of the merkle trees maintained by the
 	// persistence module that roll up into the state commitment.
-	stateTrees modules.TreeStoreModule
+	treeStore modules.TreeStoreModule
```
Consistent with https://github.com/pokt-network/pocket/pull/896#discussion_r1262253642, the persistence module should avoid holding a reference and retrieve it from the bus when needed.

bryanchriswhite commented on persistence/module.go:
```diff
@@ -103,12 +103,12 @@ func (*persistenceModule) Create(bus modules.Bus, options ...modules.ModuleOptio
 	}
 
-	treeModule, err := trees.Create(
+	treeStore, err := trees.Create(
```
This would become:

```suggestion
	_, err := trees.Create(
```

bryanchriswhite commented on persistence/module.go:
```diff
@@ -117,7 +117,7 @@ func (*persistenceModule) Create(bus modules.Bus, options ...modules.ModuleOptio
 	m.blockStore = blockStore
 	m.txIndexer = txIndexer
-	m.stateTrees = treeModule
+	m.treeStore = treeStore
```
```suggestion
```

bryanchriswhite commented on persistence/module.go:
```diff
@@ -212,7 +212,7 @@ func (m *persistenceModule) NewReadContext(height int64) (modules.PersistenceRea
 		blockStore: m.blockStore,
 		txIndexer:  m.txIndexer,
-		stateTrees: m.stateTrees,
+		stateTrees: m.treeStore,
```
(same as above - https://github.com/pokt-network/pocket/pull/896#discussion_r1262265788)

bryanchriswhite commented on persistence/trees/module.go:
```diff
@@ -8,15 +8,17 @@ import (
 
-func (*treeStore) Create(bus modules.Bus, options ...modules.TreeStoreOption) (modules.TreeStoreModule, error) {
-	m := &treeStore{}
+var _ modules.TreeStoreModule = &TreeStore{}
+
+func (*TreeStore) Create(bus modules.Bus, options ...modules.TreeStoreOption) (modules.TreeStoreModule, error) {
+	m := &TreeStore{}
+
+	bus.RegisterModule(m)
```
:100:

bryanchriswhite commented on shared/modules/treestore_module.go:
```diff
@@ -17,6 +17,7 @@ type TreeStoreFactory = FactoryWithOptions[TreeStoreModule, TreeStoreOption]
 // merkle trees that compose the state hash of pocket.
 type TreeStoreModule interface {
 	Submodule
+	TreeStoreFactory
```
`TreeStoreFactory` shouldn't need to be exported

dylanlott replied:
👍

bryanchriswhite commented on shared/modules/doc/README.md:
````diff
@@ -432,7 +434,23 @@ This is quite **important** because it unlocks a powerful concept **Dependency I
+For example, in the TreeStore code below, you can see that the TreeStoreFactory is embedded in the TreeStoreModule, which also embeds the Submodule interface.
+
+```go
+type TreeStoreFactory = FactoryWithOptions[TreeStoreModule, TreeStoreOption]
````
(same as https://github.com/pokt-network/pocket/pull/896#discussion_r1262269257)

bryanchriswhite commented on shared/modules/doc/README.md:
```diff
@@ -432,7 +434,23 @@ This is quite **important** because it unlocks a powerful concept **Dependency I
-Submodules can be registered the same way full Modules can be, by passing the Submodule to the RegisterModule function. Submodules should typically be registered to the bus for dependency injection reasons.
+Submodules can be registered the same way full Modules can be, by passing the Submodule to the RegisterModule function. Submodules should typically be registered to the bus for dependency injection reasons. Modules and submodules are all responsible for registering themselves with the Bus. This pattern emerged organically during development and is now considered best practice. Additionally, modul ...
+
+Submodule interfaces are typically defined in the `shared/modules` package with the rest of the module interfaces in a file named `XXX_module.go`, where XXX denotes the name of the submodule. That same file should contain the factory function definition for a submodule which should be embedded by the Submodule interface type. Modules should follow the same pattern but embed the Module interface i ...
```
I've been doing, and would propose, `XXX_submodule.go` for submodules.

bryanchriswhite commented on persistence/trees/module.go:
```diff
@@ -41,18 +43,21 @@ func WithLogger(logger *modules.Logger) modules.TreeStoreOption {
 
-func (t *treeStore) setupTrees() error {
-	if t.treeStoreDir == ":memory:" {
+func (t *TreeStore) GetModuleName() string { return modules.TreeStoreModuleName }
```
How do you feel about `TreeStoreSubmoduleName`?

bryanchriswhite reviewed (dismissed):
Apologies for the missing details in the docs :sweat_smile: and thanks for filling them in! :raised_hands: 

There are a few points I think we should be consistent about which I left comments/suggestions around. Overall though, looking great! :sunglasses:

h5law reviewed (changes requested):
The IBC PRs have been merged today and some of them included some changes similar to the ones here, including adding the mockgen line and bus function. It would be worth merging with main and fixing the conflicts before moving further.

h5law commented on persistence/trees/module.go:
```diff
@@ -41,18 +43,21 @@ func WithLogger(logger *modules.Logger) modules.TreeStoreOption {
-			mod.treeStoreDir = path
+		mod, ok := m.(*TreeStore)
+		if ok {
+			mod.TreeStoreDir = path
```
```suggestion
			mod.treeStoreDir = path
```
Does this need to be exported? We will always know what it's value is as we pass it in ourselves on creation

h5law commented on persistence/trees/trees.go:
```diff
@@ -72,26 +72,27 @@ type stateTree struct {
-func (t *treeStore) GetTree(name string) ([]byte, kvstore.KVStore) {
+// GetTree returns the root hash and nodeStore for the matching tree
+// stored in the TreeStore. This enables the caller to import the smt
+// and not change the one stored.
```
NIT 
```suggestion
// and not change the one stored, unless they call `Commit()` to write to the nodestore
```

dylanlott replied:
I reformatted this comment to capture this change 👍

h5law reviewed (approved):
I left a couple minor comments, but otherwise this looks good to me. Lets get it merged!

Olshansk commented on persistence/trees/trees.go:
```diff
@@ -210,24 +213,30 @@ func (t *treeStore) updateMerkleTrees(pgtx pgx.Tx, txi indexer.TxIndexer, height
-	for treeName, stateTree := range t.merkleTrees {
-		if err := stateTree.tree.Commit(); err != nil {
-			return fmt.Errorf("failed to commit %s: %w", treeName, err)
+func (t *TreeStore) Commit() error {
```
Should we leave a comment in the case that one of the trees gets committed but another doesn't?

A race condition corruption error sort of thing?

dylanlott replied:
Like a log?

Olshansk replied:
Could be a log, or a TODO comment

Are we handling this in the "atomic rollbacks" PRs later?

Eg.
```
Commit Tree #1
Commit Tree #2
Node Crashes
Did not commit Tree #3
Did not commit Tree #4
```

dylanlott replied:
Logs added for now. This function could use a multierror and return a combined error for all of the commit calls it makes instead. I'll think about this a little bit more.

Olshansk commented on persistence/trees/trees.go:
```diff
@@ -127,14 +133,11 @@ func (t *treeStore) DebugClearAll() error {
 // updateMerkleTrees updates all of the merkle trees in order defined by `numMerkleTrees`
-// * it returns the new state hash capturing the state of all the trees or an error if one occurred
-func (t *treeStore) updateMerkleTrees(pgtx pgx.Tx, txi indexer.TxIndexer, height uint64) (string, error) {
+// * it returns the new state hash capturing the state of all the trees or an error if one occurred.
```
```suggestion
// * It returns the new state hash capturing the state of all the trees or an error if one occurred.
```

Olshansk commented on persistence/trees/module_test.go:
```diff
@@ -0,0 +1,120 @@
+	persistenceMock.EXPECT().NewRWContext(int64(0)).AnyTimes()
+	persistenceMock.EXPECT().GetTxIndexer().AnyTimes()
+
+	treemod, err := trees.Create(mockBus,
+		trees.WithTreeStoreDirectory(":memory:"))
```
```suggestion
	treemod, err := trees.Create(mockBus, trees.WithTreeStoreDirectory(":memory:"))
```

Olshansk commented on persistence/trees/module_test.go:
```diff
@@ -0,0 +1,120 @@
+	treemod, err := trees.Create(mockBus,
+		trees.WithTreeStoreDirectory(":memory:"))
+	assert.NoError(t, err)
+	treemod.GetModuleName()
```
What is the purpose of this line?

Olshansk commented on persistence/trees/module_test.go:
```diff
@@ -0,0 +1,120 @@
+	persistenceMock.EXPECT().NewRWContext(int64(0)).AnyTimes()
+	persistenceMock.EXPECT().GetTxIndexer().AnyTimes()
+
+	treemod, err := trees.Create(mockBus,
```
Is there any other validation you can think of after calling create?

E.g. the default root hash
E.g. Calling some of the exposed functions

Olshansk commented on persistence/trees/module_test.go:
```diff
@@ -0,0 +1,120 @@
+		m.SetBus(mockBus)
+	}).AnyTimes()
+	mockModulesRegistry := mockModules.NewMockModulesRegistry(ctrl)
+	mockModulesRegistry.EXPECT().GetModule(peerstore_provider.PeerstoreProviderSubmoduleName).Return(nil, runtime.ErrModuleNotRegistered(peerstore_provider.PeerstoreProviderSubmoduleName)).AnyTimes()
```
NIT: I think multilining these more complex mocks really makes it much easier to understand what's being called & returned.

Ditto elsewhere.

```suggestion
mockModulesRegistry.EXPECT().
	GetModule(peerstore_provider.PeerstoreProviderSubmoduleName).
	Return(nil, runtime.ErrModuleNotRegistered(peerstore_provider.PeerstoreProviderSubmoduleName)).
	AnyTimes()
```

Olshansk reviewed (changes requested):
PTAL at https://github.com/pokt-network/pocket/pull/896/files#r1268746846 as well

dylanlott commented:
> PTAL at [#896 (files)](https://github.com/pokt-network/pocket/pull/896/files#r1268746846) as well

Okay, I believe I addressed all comments 👍

h5law commented on persistence/trees/module_test.go:
```diff
@@ -0,0 +1,176 @@
+
+	// root hash should be empty for empty tree
+	root, ns := treemod.GetTree(trees.TransactionsTreeName)
+	require.Equal(t, root, []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
```
```suggestion
	require.Equal(t, root, make([]byte, 32))
```

Olshansk commented on persistence/trees/module.go:
```diff
@@ -8,15 +8,17 @@ import (
 func (*treeStore) Create(bus modules.Bus, options ...modules.TreeStoreOption) (modules.TreeStoreModule, error) {
 	m := &treeStore{}
 
+	bus.RegisterModule(m)
```
I did a search for `for _, option := range options {` in the codebase and found that we always do `RegisterModule` AFTER applying the options.

I know it's unlikely that the options will impact registration (maybe with the exception of tests/mocks) but just want to stay consistent.

dylanlott replied:
Noted! I will add this to the module documentation 👍

Olshansk reviewed (approved):
Approving with one minor change request. PTAL

Olshansk commented on persistence/trees/trees.go:
```diff
@@ -210,24 +213,34 @@ func (t *treeStore) updateMerkleTrees(pgtx pgx.Tx, txi indexer.TxIndexer, height
+
+	for name, treeStore := range t.merkleTrees {
+		if err := treeStore.tree.Commit(); err != nil {
+			t.logger.Err(err).Msgf("failed to commit to %s tree: changes will not be saved - this should be investigated", name)
```
```suggestion
			t.logger.Err(err).Msgf("TECHDEBT: failed to commit to %s tree: changes will not be saved - this should be investigated", name)
```

Olshansk commented on persistence/trees/trees.go:
```diff
@@ -210,24 +213,34 @@ func (t *treeStore) updateMerkleTrees(pgtx pgx.Tx, txi indexer.TxIndexer, height
+// Commit commits changes in the sub-trees to the root tree and then commits updates for each sub-tree.
+func (t *treeStore) Commit() error {
+	if err := t.rootTree.tree.Commit(); err != nil {
+		t.logger.Err(err).Msg("failed to commit root tree: changes to sub-trees will not be committed - this should be investigated")
```
```suggestion
		t.logger.Err(err).Msg("TECHDEBT: failed to commit root tree: changes to sub-trees will not be committed - this should be investigated")
```

Olshansk replied:
I think a techdebt is important here to make sure we don't forget about it

Olshansk commented on shared/modules/doc/README.md:
```diff
@@ -434,7 +434,20 @@ This is quite **important** because it unlocks a powerful concept **Dependency I
 This is useful not only for prototyping but also for different use cases such as the `p1` CLI and the `pocket` binary where different implementations of the same module are necessary due to the fact that the `p1` CLI doesn't have a persistence module but still needs to know what's going on in the network.
 
-**Registration**: Submodules can be registered the same way full Modules by passing the Submodule to the `RegisterModule` function. Submodules should typically be registered to the bus for dependency injection reasons.
+**Registration**: Submodules can be registered the same way full Modules by passing the Submodule to the `RegisterModule` function. Submodules should typically be registered to the bus for dependency injection reasons. Registration should occur _after_ processing the module's options.
```
Thank you for adding this!
